### DISCUSSION TITLE
How does the matched filter work?

### DISCUSSION THREAD

**Discussion by mapfiable** [sep-developers/sep#91] [Q&A] [answered]
In the [SE user manual](https://astromatic.net/pubsvn/software/sextractor/trunk/doc/sextractor.pdf) (p. 18, chap. 7.2.1), they already derive a matched filter. In the [SEP docs](https://sep.readthedocs.io/en/v1.1.x/filter.html) it says that this is based on the assumption that noise is equal across the image (I haven't found a reference to this in the SE manual, but I guess it's implied by how the filter is derived; only afterwards do they assume white noise). In this case, the filter appearently simplifies to a regular convolution of the image with the filter kernel, which equates to flipping the kernel on the x and y axis and then multiplying each value in the kernel with the corresponding one in the image and and summing up the results. It also says as much in the [wiki article about convolution](https://en.wikipedia.org/wiki/Kernel_(image_processing)).

The SEP solution for the matched filter in case of independent noise in each pixel seems more complicated than simply flipping the filter on its x and y axes. I asked for the visual representation of the application on [stackexchange](https://dsp.stackexchange.com/q/70379/52326), but there I was told all you do is flip the filter. So I am confused, but I guess they assumed the equal noise case. What does the application of the matched filter actually look like (e.g. analogous to how convolution is displayed on wiki)?

I really like your derivation using matrices, since all the sources on matched filters I managed to find deal with sonar / radar applications in 1D and use itegrals and such. Can you point me to a source using matrix notation (other than the [wiki article](https://en.wikipedia.org/wiki/Matched_filter))? I didn't manage to find one.

What is the covariance matrix in the derivation? Is this where the RMS-map is used?

**Comment by kboone** [comment, marked as answer]
The derivation of the matched filter formula can be found on the [sep readthedocs page](https://sep.readthedocs.io/en/stable/filter.html) or on [wikipedia](https://en.wikipedia.org/wiki/Matched_filter#Derivation). The final filter that we use in sep is the second to last equation on the sep readthedocs page. If the uncertainty is the same for each pixel, then this formula simplifies to the one that you are referring to, i.e. a convolution with the kernel.

The derived equation is valid for any covariance matrix, although we currently only support covariance matrices in `sep` where all of the pixels have independent noise (i.e. the covariance matrix only has terms on the diagonal). In source extractor, they make a further assumption that the noise for each pixel is the same. This is oftentimes what you want, but not always. I implemented the version that is currently in `sep` to handle stacked images where you can have many stacked exposures in some places in the image (i.e. low noise) and few exposures in other places (i.e. higher noise).

To use the matched filter in `sep`, you simply manually specify the errors for each of your pixels with the `err` keyword to `sep.extract`.

**Reply by mapfiable** [reply]
Hi, thanks! I was wondering if you know any textbook or paper where the matched filter is derived; basically, I'd like to know its source. If you search the internet for "matched filter", all you find is stuff about radar and sonar, but nothing about the application to images. And it's usually all about the equal noise case. 

As far as I understand, the `err` map is used for the thresholding, but does not play any role in the matched filter, correct? So how are the convolution matrices created? What data plays a role in it?

**Reply by kboone** [reply]
I'm not sure what the original source is. Matched filters have been around for a very long time. I derived the formula that is used in `sep` myself (although a similar derivation can be found other places).

The matched filter requires an error map. This can be either a known error map if you have it from say the data reduction pipeline, or the one that is estimated from `sep.Background`. The error map is used with the matched filter to calculate the signal-to-noise of an object being present at each pixel in the image. The `thresh` keyword of `sep.extract` only keeps objects with a signal-to-noise above some threshold.

A matched filter will find objects that look like whatever filter you use. By default, sep uses a 3x3 filter given by `[[1,2,1], [2,4,2], [1,2,1]]`. You can optimize this for different science cases. For example, if you are looking for stars in an image, the optimal filter is one that looks like a star in the image, i.e. the PSF of the image. In practice, the simple kernel is usually fine.

**Reply by mapfiable** [reply]
> I'm not sure what the original source is. Matched filters have been around for a very long time. I derived the formula that is used in `sep` myself (although a similar derivation can be found other places).

Oh wow! I guess I can just refer to your derivation then. But you did learn about it before, did you not? Maybe I will still do some digging and try to find the original source.

Thanks!

**Reply by kboone** [reply]
I started from the derivation in source extractor. Matched filters have been around for a very long time, so I don't know what the "original source" is.
